pkg/controller/v1beta1/trainedmodel/controller.go

```diff
@@ -0,0 +1,100 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1beta1api "github.com/kubeflow/kfserving/pkg/apis/serving/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TrainedModelReconciler reconciles a TrainedModel object
+type TrainedModelReconciler struct {
+	client.Client
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=serving.kubeflow.org,resources=trainedmodel,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=serving.kubeflow.org,resources=trainedmodel/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=serving.knative.dev,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=serving.knative.dev,resources=services/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=serviceaccounts,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=namespaces,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;update;patch;delete
+func (r *TrainedModelReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("trainedmodel", req.NamespacedName)
+
+	// Fetch the TrainedModel instance
+	trainedModel := &v1beta1api.TrainedModel{}
+	if err := r.Get(context.TODO(), req.NamespacedName, trainedModel); err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	log.Info("Reconciling TrainedModel", "apiVersion", trainedModel.APIVersion, "trainedModel", trainedModel.Spec)
+
+	shardManager := shard.ShardManager{Strategy: shard.Memory}
+	shardId := shardManager.GetShardIdForTrainedModel(trainedModel)
+	// Use trainedModel's parent InferenceService field to get the multi-model configMap
+	multiModelConfigMapName := constants.MultiModelConfigMapName(trainedModel.Spec.InferenceService, shardId)
+	configMap := &v1.ConfigMap{}
+	err := r.Get(context.TODO(), types.NamespacedName{Name: multiModelConfigMapName, Namespace: req.Namespace}, configMap)
+	if err != nil {
+		log.Error(err, "Failed to find Multi-model ConfigMap", "name", multiModelConfigMapName, "namespace", req.Namespace)
+		// Error reading the object - requeue the request.
+		return reconcile.Result{Requeue: true, RequeueAfter: 10 * time.Second}, err
+	}
+
+	//Reconcile multi-model configmap to add/update/remove models
+	configMapReconciler := multimodelconfig.NewConfigMapReconciler(r.Client, r.Scheme)
```

Comment:
We shouldn't be reinstantiating this within a reconcile loop. Can you make this a member of the TrainedModelReconciler struct and initialize it once?